### PR TITLE
Pin test dependencies

### PR DESCRIPTION
**Pull request by Czaki**
# Description

In this PR I add constraints files for napari test dependencies and workflow to perform every week update of dependencies. 

The motivation for this PR is to avoid regular failures not related to PR context but caused by some dependency release. 

Also, this PR will provide dependency versions on which napari works during release time.

This PR is inspired by workflow from PartSeg that example output could be found here https://github.com/4DNucleome/PartSeg/pull/932 (but here is not listening all updated dependencies) 

This PR will require adding `GHA_TOKEN` to repository secrets with the following permissions: 

obraz

# References

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

File: .github/workflows/upgrade_test_constrains.yml
Comment (jaimergp) [inline .github/workflows/upgrade_test_constrains.yml:1]: Typo on the filename I think? `..._constraints.yml`

File: tools/constraints/version_blacklist.txt
Comment (jaimergp) [inline tools/constraints/version_blacklist.txt:1]: Consider using `version_denylist.txt` instead?

File: .github/workflows/upgrade_test_constrains.yml
Comment (jaimergp) [inline .github/workflows/upgrade_test_constrains.yml]:
```suggestion
name: Upgrade test constraints
```

**Comment by jaimergp** [comment]
I like this! I recently learnt about the [Dependency submissions API](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/using-the-dependency-submission-api), which might be related to this in terms of purpose. Having the constraints already there should make it super easy to integrate!

File: tools/constraints/version_blacklist.txt
Comment (Czaki) [reply tools/constraints/version_blacklist.txt:1]: Yes. It is much better

File: .github/workflows/upgrade_test_constrains.yml
Comment (Czaki) [reply .github/workflows/upgrade_test_constrains.yml]: Thanks

File: .github/workflows/upgrade_test_constrains.yml
Comment (Czaki) [reply .github/workflows/upgrade_test_constrains.yml:1]: No spellcheck during create file :(

**Comment by jni** [comment]
@Czaki can you elaborate on what kind of token I need to create? Is this a personal access token or something else? Detailed instructions appreciated. 😊

File: .github/workflows/upgrade_test_constraints.yml
Comment (jni) [inline .github/workflows/upgrade_test_constraints.yml:47]: @Czaki could you add a long comment here to explain what these lines and each of the options do? They are very long and scary. 😅

File: .github/workflows/upgrade_test_constraints.yml
Comment (jni) [inline .github/workflows/upgrade_test_constraints.yml]:
```suggestion
            This PR is automatically created and updated by this napari GitHub Action:
            https://github.com/napari/napari/tree/main/.github/workflows/upgrade_test_contraints.yml
            It ensures that dependencies are tested with the most recent version.
```

File: tools/check_updated_packages.py
Comment (jni) [inline tools/check_updated_packages.py]:
```suggestion
    print("only indirect updates")
```

**Review by jni — COMMENTED** [review]
@Czaki thanks for this! Almost ready I think!

File: .github/workflows/upgrade_test_constraints.yml
Comment (Czaki) [reply .github/workflows/upgrade_test_constraints.yml:47]: * `python3.8 -m piptools compile` - call `pip-compile` but ensure proper intepretter 
* `--upgrade` upgrade to the latest possible version. Without this `pip-compile` will take a look to output files and reuse versions (so will ad something on when adding dependency.
* `-o resources/constraints/constraints_py3.8.txt` - output file
* `setup.cfg resources/constraints/version_denylist.txt` - source files. the `resources/constraints/version_denylist.txt - contains our test specific constraints like `pytes-cov` 
* `--extra pyqt5` etc - names of extra sections from `setup.cfg` that should be checked for the dependencies list (maybe we could create a super extra section to collect them all in) 
* `--allow-unsafe` - allow to put in constraints things like setuptools (look at the end of one of the generated files). It will be the default behavior in the future. 
* `--strip-extras` - pip constrains format does not allow to specify extras (like `dask[array]`) so we need to use this option
* `--resolver=backtracking` - future default resolver. It fasters. Lower probability of long CI run. 

Did you want to put this comment in yml file?

Comment (jni) [reply .github/workflows/upgrade_test_constraints.yml:47]: It sounds kind of ridiculous, yes, but kind of yes? I wonder if it would work with a multiline string with trailing yaml comments?

**Comment by Czaki** [comment]
> @Czaki can you elaborate on what kind of token I need to create? Is this a personal access token or something else? Detailed instructions appreciated. blush

Go to napari Settings > Third-party Access > Personal access tokens > Settings 
image

Then allow Fine-grained personal access tokens but keep "Require administrator approval"

image


Go to personal Settings > Developer Settings > Personal access tokens > Fine-grained tokens
image

Generate new Token

Set resource owner to napari. Set long lime of expiration. Use  "Only select repositories"

Here is an example of the configuration for PartSeg, as the napari repository do not allow currently for these tokens:
image

Then in repository permissions, set Pull request permissions to Read and Write:
image

Click "generate token". 

Such toke could also solve the problem of not starting CI on updating vendor dependency PR.

**Comment by jni** [comment]
Ok and then the value of the token I put into a new repository secret as GHA_TOKEN, yes? I just did that. Hopefully I didn't mess up any steps... 😬 🤞

**Comment by Czaki** [comment]
So to test this, we need to merge and trigger this workflow manually.

**Review by jni — APPROVED** [review]

**Comment by jni** [comment]
Could you mention in Zulip #reviews that this is ready? We should wait 24h to make sure ~everyone is on board, as it's a significant change to our workflow.

**Comment by Czaki** [comment]
I have added such a message.

**Comment by brisvag** [comment]
Copying from zulip:
```quote
can you summarize how this can have a "big impact to our workflow"?
```
```quote
The test depenecies will be pinned base on constrains files, but every week
(with option to manual trigger) there will be PR with update of all test
dependencies constrains.
So it should reduce number of extinguishing fires, but increase maintenance.
So it will be nice to more dev understand what happens there if there happen
need to update configuration.
```

**Comment by brisvag** [comment]
Seems reasonable; I guess we can always revert if it turns out to be too much burden.

**Review by brisvag — APPROVED** [review]

**Comment by psobolewskiPhD** [comment]
Am I getting this right:
The workflow will open a PR with bumped dependencies, if there are any.
In that PR, tests with **these new ones** will run, right?
So if they pass, we can basically merge safely.
If tests fail, we know we need to keep the pin and/or solve the issue.

**Comment by Czaki** [comment]
Yes. The whole setup token step is to allow running tests for automatically opened PR.

**Comment by kephale** [comment]
@Czaki out of curiosity, why are the pip install and examples tests using 3.9 instead of 3.10?

**Comment by kephale** [comment]
@Czaki is `tools/constraints/version_blacklist.txt` missing?

**Comment by Czaki** [comment]
> @Czaki is `tools/constraints/version_blacklist.txt` missing?

Was renamed to `version_denylist`. I do not regenerate files to simpler check if its working good after merge. 



> @Czaki out of curiosity, why are the pip install and examples tests using 3.9 instead of 3.10?

At some moment there was the decision to not use the latest supported, but one before. When we enable python 3.11 test then we should switch it to python 3.10

**Comment by kephale** [comment]
Awesome, thank you @Czaki !

**Comment by Czaki** [comment]
@Carreau This PR may conflict with parts of your work. What did you think about it?

**Comment by alisterburt** [comment]
Love this - going to merge because we have two approving reviews but we can always revert if @Carreau has issues

**Pull request merged**
